src/operator/numpy/random/np_location_scale_op.h

```diff
@@ -0,0 +1,510 @@
+                 const Shape<ndim> &stride, const Shape<ndim> &oshape,
+                 const Tensor<xpu, 1, IType>& array, float scalar,
+                 const Tensor<xpu, 1, float>& noise, const Tensor<xpu, 1, OType>& out) {
+      Kernel<gumbel_one_scalar_kernel<ndim, IType, OType>, xpu>::Launch(
+        s, out_size, scalar_pos, stride, oshape, array.dptr_, scalar, noise.dptr_, out.dptr_);
+  }
+};
+
+template <int ndim, typename IType, typename OType>
+struct logistic_kernel {
+  MSHADOW_XINLINE static void Map(index_t i, const Shape<ndim> &lstride,
+                                  const Shape<ndim> &hstride,
+                                  const Shape<ndim> &oshape, IType *loc,
+                                  IType *scale, float *noise, OType *out) {
+    Shape<ndim> coord = unravel(i, oshape);
+    auto lidx = static_cast<index_t>(dot(coord, lstride));
+    auto hidx = static_cast<index_t>(dot(coord, hstride));
+    IType loc_value = loc[lidx];
+    IType scale_value = scale[hidx];
+    noise[i] = log(noise[i]) - log(1 - noise[i]);
+    out[i] = loc_value + noise[i] * scale_value;
+  }
+};
+
+struct logistic_two_tensor {
+  template <typename xpu, int ndim, typename IType, typename OType>
+  static void op(Stream<xpu> *s, const int out_size, const Shape<ndim> &lstride,
+                 const Shape<ndim> &hstride, const Shape<ndim> &oshape,
+                 const Tensor<xpu, 1, IType>& loc, const Tensor<xpu, 1, IType>& scale,
+                 const Tensor<xpu, 1, float>& noise, const Tensor<xpu, 1, OType>& out) {
+      Kernel<logistic_kernel<ndim, IType, OType>, xpu>::Launch(
+        s, out_size, lstride, hstride, oshape, loc.dptr_,
+        scale.dptr_, noise.dptr_, out.dptr_);
+  }
+};
+
+template <int ndim, typename IType, typename OType>
+struct gumbel_kernel {
+  MSHADOW_XINLINE static void Map(index_t i, const Shape<ndim> &lstride,
+                                  const Shape<ndim> &hstride,
+                                  const Shape<ndim> &oshape, IType *loc,
+                                  IType *scale, float *noise, OType *out) {
+    Shape<ndim> coord = unravel(i, oshape);
+    auto lidx = static_cast<index_t>(dot(coord, lstride));
+    auto hidx = static_cast<index_t>(dot(coord, hstride));
+    IType loc_value = loc[lidx];
+    IType scale_value = scale[hidx];
+    noise[i] = -log(-log(noise[i]));
+    out[i] = loc_value + noise[i] * scale_value;
+  }
+};
+
+struct gumbel_two_tensor {
+  template <typename xpu, int ndim, typename IType, typename OType>
+  static void op(Stream<xpu> *s, const int out_size, const Shape<ndim> &lstride,
+                 const Shape<ndim> &hstride, const Shape<ndim> &oshape,
+                 const Tensor<xpu, 1, IType>& loc, const Tensor<xpu, 1, IType>& scale,
+                 const Tensor<xpu, 1, float>& noise, const Tensor<xpu, 1, OType>& out) {
+    Kernel<gumbel_kernel<ndim, IType, OType>, xpu>::Launch(
```

Comment:
Why wrapping a a kernel inside another kernel?

Reply:
For <ndim, IType, OType> could not be got during the registration.